### PR TITLE
#382/Fix UI lag when modifying scenario fields. 

### LINKED ISSUE #382: [IMPORTANT] Fix delay when adding mitigation intervals
# 🐛 Bug Report

## How to reproduce

Steps to reproduce the issue:

 1. Open the application in a browser
 2. Click "Add" button in "Mitigation"
 3. Observe that it takes a while for the new interval to appear in the table

## 😯 Current Behavior

New intervals appear after delay

## 🤔 Expected Behavior

New intervals should appear instantly

## 🔦 Context

It's either "Add" button is wrongly debounced or it waits for the algorithm to finish, or both.

Importance:  Intervals is a commonly asked for feature. Should be included into the next release.

## 🌍 Your Environment

Any

## Related

 - https://github.com/neherlab/covid19_scenarios/pull/323
 - https://github.com/neherlab/covid19_scenarios/issues/14

### PR DESCRIPTION
**Pull request by abrie**
## Related issues and PRs

Fixes #382, similar to #402, #403 
 
## Description

This PR attempts to reduces UI lag when editing Scenario parameters. The lag is primarily due to the heavy `DeterministicLinePlot` component which has several complex state dependencies. The current codebase mitigates the lag by adding a 1000msec debounce to the `setCustomScenario()`. That works – except it introduces a new lag (of 1000ms) to the mitigation interval editor. #402 and #403 fix the problem by wrapping the DeterministicLinePlot with `React.memo()`, but this may have unintended consequences because of the complex state.

This PR removes the debounce from `setCustomScenario`, and moves it to ~~the render method of~~ a wrapper component around  `DeterministicLinePlot`. The net effect is the same, and we avoid the potentially troublesome memoization.

~~Drawback is that the DeterministicLinePlot is re-rendered every 1000ms; but if there is no change then the cost is minimal.~~ (no longer an issue)

~~**Another drawback is that the graph interactions are now sluggish**~~ (no longer an issue)

## Impacted Areas in the application

- General UI input responsiveness
- (re)rendering rate of Results Card, particularly DeterministicLinePlot.

## Testing

Test the fix by looking for UI sluggishness. The scenario inputs and mitigation editor should provide snappy feedback. Test with autorun enabled and disabled. Verify that tooltips are shown as expected when hovering over the charts.

add-mitigation-interval

### CODE REVIEW

File: src/components/Main/Main.tsx [hunk 1]
```diff
@@ -163,7 +163,7 @@ function Main() {
       const mitigationIntervals = _.map(newParams.containment.mitigationIntervals, _.cloneDeep)
       scenarioDispatch(setContainmentData({ data: { mitigationIntervals } }))
     }
-  }, 1000)
+  }, 0)
```
Comment (abrie) [inline src/components/Main/Main.tsx:166]: Maybe just remove the debounce altogether...

File: src/components/Main/Results/DeterministicLinePlot.tsx [hunk 2]
```diff
@@ -109,6 +111,9 @@ export function DeterministicLinePlot({
   const { t } = useTranslation()
   const chartRef = React.useRef(null)
   const [enabledPlots, setEnabledPlots] = useState(Object.values(DATA_POINTS))
+  const [rendered, setRendered] = useState(<></>)
```
Comment (abrie) [inline src/components/Main/Results/DeterministicLinePlot.tsx:114]: We cache the render output as state. Initially it contains nothing.

File: src/components/Main/Results/DeterministicLinePlot.tsx [hunk 3]
```diff
@@ -109,6 +111,9 @@ export function DeterministicLinePlot({
...
+
+  const [debouncedRender] = useDebouncedCallback(() => setRendered(render()), 1000)
```
Comment (abrie) [inline src/components/Main/Results/DeterministicLinePlot.tsx:116]: Store the rendered output into the render state. The render function can take upwards of 60msec to finish, so we debounce to a maximum of once every 1000 msec.

File: src/components/Main/Results/DeterministicLinePlot.tsx [hunk 4]
```diff
@@ -248,103 +253,111 @@ export function DeterministicLinePlot({
...
-
-                <Legend
-                  verticalAlign="top"
-                  formatter={(v, e) => legendFormatter(enabledPlots, v, e)}
-                  onClick={(e) => {
-                    const plots = enabledPlots.slice(0)
-                    enabledPlots.includes(e.dataKey) ? plots.splice(plots.indexOf(e.dataKey), 1) : plots.push(e.dataKey)
-                    setEnabledPlots(plots)
+  function render() {
+    return (
```
Comment (abrie) [inline src/components/Main/Results/DeterministicLinePlot.tsx:257]: Here the original component's 'return' statement is wrapped into a function which will be debounced.

File: src/components/Main/Results/DeterministicLinePlot.tsx [hunk 5]
```diff
@@ -248,103 +253,111 @@ export function DeterministicLinePlot({
...
+            )
+          }}
+        </ReactResizeDetector>
+      </div>
+    )
+  }
+
+  debouncedRender()
+
+  return rendered
```
Comment (abrie) [inline src/components/Main/Results/DeterministicLinePlot.tsx:362]: Invoked the debounced render to (maybe) update the render state. Then return the (maybe updated) render state.

**Comment by gj262** [comment]
This is kinda cool. I would never have thought of debouncing render(). It fixes the main culprit of the problem.

**Comment by abrie** [comment]
> This is kinda cool. I would never have thought of debouncing render(). It fixes the main culprit of the problem.

There are some drawbacks; one being that the graph interaction is now sluggish. But perhaps there is a clean workaround. Worth a bit more work.

And, unfortunately this fix is not compatible with the the merge of #378.

**Comment by gj262** [comment]
And the tooltips disappear if the mouse sits on the graph.

**Comment by abrie** [comment]
Ok, got some fixes for these issues: https://github.com/neherlab/covid19_scenarios/pull/412#issuecomment-609518115 and https://github.com/neherlab/covid19_scenarios/pull/412#issuecomment-609517558.

The issues stem from the fact that there are two update scenarios for the component. Either its modified through props externally, or re-rendered from within through zoom/tooltip actions. The internal actions should not be debounced, and external actions must be debounced. So this fix moves the debouncing logic out of `DeterministicLinePlot` and into a wrapper class component (`DebouncedDeterministicLinePlot`). The wrapper takes care of debouncing changes to the props, while leaves internal events untouched.

File: src/components/Main/Results/DeterministicLinePlot.tsx [hunk 6]
```diff
@@ -97,6 +97,24 @@ function legendFormatter(enabledPlots: string[], value: string, entry: any) {
...
+  shouldComponentUpdate() {
+    this.debouncedRender()
+    return false;
+  }
+
+  render() {
+    return <DeterministicLinePlot {...this.props}/>
+  }
+}
+
```
Comment (abrie) [inline src/components/Main/Results/DeterministicLinePlot.tsx:117]: The goal was a functional solution, but `DeterministicLinePlot` has a lot of state to wrangle. This class intercepts props changes coming from outside the component and debounces the update to 1000msec intervals.

File: src/components/Main/Results/ResultsCard.tsx [hunk 7]
```diff
@@ -191,7 +191,7 @@ function ResultsCardFunction({
         </Row>
         <Row noGutters>
           <Col>
-            <DeterministicLinePlot
+            <DebouncedDeterministicLinePlot
```
Comment (abrie) [inline src/components/Main/Results/ResultsCard.tsx:194]: Use the debounced version.

File: src/components/Main/Main.tsx [hunk 8]
```diff
@@ -130,7 +130,7 @@ function Main() {
     }
   }, [autorunSimulation, debouncedRun, scenarioState, locationSearch, severity])
 
-  const [setScenarioToCustom] = useDebouncedCallback((newParams: AllParams) => {
+  const setScenarioToCustom = (newParams: AllParams) => {
```
Comment (abrie) [inline src/components/Main/Main.tsx:133]: Remove the useDebounceCallback here because the debounce is now handled by `DebouncedDeterministicLinePlot`.

**Comment by abrie** [comment]
#437

**Pull request closed without merging**